### PR TITLE
Added the --temps-dir option

### LINKED ISSUE #10971: Names of intermediate files collide when running rustc in parallel
For example, given an empty lib.rs, this happens:

```
rustc --lib lib.rs & rustc --rlib lib.rs
error: linking with `cc` failed: exit code: 1
note: cc arguments: '-m64' '-L/usr/local/lib/rustc/x86_64-unknown-linux-gnu/lib' '-o' 'liblib-9ed81b85-0.0.so' 'lib.o' 'lib.metadata.o' '-Wl,--as-needed' '-L/tmp/.rust' '-L/tmp' '-L/usr/local/lib/rustc/x86_64-unknown-linux-gnu/lib' '-lstd-04ff901e-0.9-pre' '-lrt' '-ldl' '-lm' '-lpthread' '-lstdc++' '-shared' '-lmorestack' '-Wl,-rpath,$ORIGIN/../usr/local/lib/rustc/x86_64-unknown-linux-gnu/lib' '-Wl,-rpath,/usr/local/lib/rustc/x86_64-unknown-linux-gnu/lib'
note: /usr/bin/ld: reopening lib.o: No such file or directory

/usr/bin/ld: final link failed: No such file or directory
collect2: error: ld returned 1 exit status

error: aborting due to previous error
task 'rustc' failed at 'explicit failure', /home/siege/src/rust2/src/libsyntax/diagnostic.rs:102
task '<main>' failed at 'explicit failure', /home/siege/src/rust2/src/librustc/lib.rs:398
```

This could be fixed in two main ways.
1. Add an option (e.g. `--temps-path`) to store the intermediates in a specific location
2. Add some kind of hash to the names of the intermediates to stop them from colliding. Can't just be the pkgid hash, as it'd still fail the above test.

### PR DESCRIPTION
**Pull request by torhovland**
Fixes #10971.

The new `--temps-dir` option puts intermediate files in a user-specified directory. This provides a fix for the issue where parallel invocations of rustc would overwrite each other's intermediate files.

No files are kept in the intermediate directory unless `-C save-temps=yes`.

If additional files are specifically requested using `--emit asm,llvm-bc,llvm-ir,obj,metadata,link,dep-info,mir`, these will be put in the output directory rather than the intermediate directory.

This is a backward-compatible change, i.e. if `--temps-dir` is not specified, the behavior is the same as before.

### CODE REVIEW

**Review by davidtwco — APPROVED** [review]
Implementation looks good to me. Could you submit a [major change proposal](https://forge.rust-lang.org/compiler/mcp.html)? The compiler team's [process for adding new `rustc` options](https://forge.rust-lang.org/compiler/new_option.html#proposal-and-check-off) requires it - just to make sure that it's something we want to support going forward :)

**Comment by davidtwco** [comment]
Apologies, this slipped past me after the major change proposal was accepted. Thanks for your patience.

@bors r+

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/86373#issuecomment-862672904
@bors r-

**Comment by crlf0710** [comment]
@torhovland Ping from triage! Seems this pr's change doesn't work as expected on some platform. Do you want to investigate? Thanks!

**Comment by torhovland** [comment]
Not sure, but it could be that this test command doesn't run as expected on Windows:

```
$(RUSTC) --crate-type=lib --temps-dir=$(TMPDIR)/temp1 --out-dir=$(TMPDIR) $(TMPDIR)/lib.rs \
		& $(RUSTC) --crate-type=cdylib --temps-dir=$(TMPDIR)/temp2 --out-dir=$(TMPDIR) $(TMPDIR)/lib.rs
```

https://github.com/rust-lang/rust/pull/86373/files#diff-03a597d35d09133542008990925ff15740cc1781c0bf2c82d21af21712ac1eff

**Comment by crlf0710** [comment]
@torhovland Ping from triage, any updates?

**Comment by crlf0710** [comment]
@torhovland Ping from triage, any updates?

**Comment by camelid** [comment]
Is this flag supposed to be initially unstable or does the MCP mean that it's insta-stable (and thus should also be labeled with `relnotes`)?

**Comment by torhovland** [comment]
Sorry, I haven't looked at this after that Windows test failed, but I or somebody else should investigate the test command I mention above on Windows.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@torhovland can you address the merge conflicts and set `S-waiting-on-review` ?

**Comment by torhovland** [comment]
I've rebased and changed the test command so it's not split over two lines. This may or may not help the test succeed on Windows. It would be nice if you could try merging it, so we at least get a new test output, because the old logs have expired. I suggest not rolling up this PR, as there's a good chance it will fail again.

**Comment by torhovland** [comment]
@davidtwco Note that this PR has already been reviewed, and the only change after is a rebase and a trivial change to a test. It would be nice to get another run of the failing test.

By the way, if there is a way I can run the test suite (or even better, just the single test) on the failing platform (i686-msvc-1), let me know.

**Comment by davidtwco** [comment]
@bors r+

> By the way, if there is a way I can run the test suite (or even better, just the single test) on the failing platform (i686-msvc-1), let me know.

I think you would need to be on the failing platform, but then you can just run the test suite as normal, `./x.py test src/test/run-make --test-args issue-10971-temps-dir`.

**Comment by torhovland** [comment]
> I think you would need to be on the failing platform, but then you can just run the test suite as normal, `./x.py test src/test/run-make --test-args issue-10971-temps-dir`.

Hmm, yeah, I suppose I can try building it all on that platform. Will look into it if this fails again.

**Comment by bjorn3** [comment]
You should edit `src/ci/github-actions/ci.yml` and then run `./x.py run src/tools/expand-yaml-anchors` to update `.github/workflows/ci.yml`.

**Comment by torhovland** [comment]
> You should edit `src/ci/github-actions/ci.yml` and then run `./x.py run src/tools/expand-yaml-anchors` to update `.github/workflows/ci.yml`.

Ok, sorry, didn't know that. Anyway, I'm not really going to change any of that, I was just trying to run the Windows tests on my own fork so I could figure out why they're failing.

**Comment by torhovland** [comment]
@davidtwco All right, I was able to reproduce this on Windows and fix the test. All that's really changed since the review is the missing `include` in https://github.com/rust-lang/rust/pull/83846/commits/d4bcee9638cf8fdf7e9297438e70e20b212c6535.

I believe this should finally pass now, so feel free to rollup.

**Comment by davidtwco** [comment]
> Also, there's a question about release notes from @camelid above. Let me know if I need to add anything.

Just to be on the safe side, could you make this option require `-Z unstable-options` for just now, and then we can merge it and look into stabilisation shortly after.

File: src/doc/rustc/src/command-line-arguments.md [hunk 1]
```diff
@@ -194,15 +194,6 @@ This flag controls the output filename.
...
 the [`-o` flag](#option-o-output) is used.
 
-<a id="option-temps-dir"></a>
-## `--temps-dir`: directory to write the intermediate files in
-
-Intermediate files will be written to this directory. If not set, the output
-directory is used. This option is useful if you are running more than one
-instance of `rustc` (e.g. with different `--crate-type` settings), and you
-need to make sure they are not overwriting each other's intermediate files.
-No files are kept unless `-C save-temps=yes` is also set.
```
Comment (camelid) [inline src/doc/rustc/src/command-line-arguments.md]: You could move this to the Unstable Book (in `src/doc/unstable-book`) instead.

File: compiler/rustc_interface/src/tests.rs [hunk 2]
```diff
@@ -772,6 +772,7 @@ fn test_debugging_options_tracking_hash() {
     tracked!(src_hash_algorithm, Some(SourceFileHashAlgorithm::Sha1));
     tracked!(symbol_mangling_version, Some(SymbolManglingVersion::V0));
     tracked!(teach, true);
+    tracked!(temps_dir, String::from("abc"));
```
Comment (camelid) [inline compiler/rustc_interface/src/tests.rs:775]: I think you need to wrap the second argument with `Some`.

**Comment by torhovland** [comment]
> Just to be on the safe side, could you make this option require `-Z unstable-options` for just now, and then we can merge it and look into stabilisation shortly after.

OK, done!

**Comment by torhovland** [comment]
Well, this is becoming a bit of a saga. The Windows tests should run fine, and they also run fine on my personal `x86_64-unknown-linux-gnu`. But on `dist-various-1`, which is also `x86_64-unknown-linux-gnu` it fails now, because it has added `-Clinker='arm-none-eabi-gcc'`, and apparently that linker doesn't understand `-m64`.

Some other tests do `unexport RUSTC_LINKER`. My best bet right now is to try that.

**Comment by torhovland** [comment]
@davidtwco This needs another full test run.

**Comment by torhovland** [comment]
@davidtwco I was able to reproduce this test failure locally, and verify that it works better with staticlib rather than cdylib. Can you please give this another go?

**Pull request merged**